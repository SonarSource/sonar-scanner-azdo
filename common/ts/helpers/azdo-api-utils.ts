import * as tl from "azure-pipelines-task-lib/task";
import * as vm from "azure-devops-node-api";
import {
  JsonPatchDocument,
  JsonPatchOperation,
  Operation,
} from "azure-devops-node-api/interfaces/common/VSSInterfaces";

export interface IPropertyBag {
  propertyName: string;
  propertyValue: string;
}

export async function addBuildProperty(properties: IPropertyBag[]) {
  const collectionUri = tl.getVariable("System.TeamFoundationCollectionUri") + "/";
  const teamProjectId = tl.getVariable("System.TeamProjectId") + "/";
  const buildId = tl.getVariable("Build.BuildId");

  const patchBody: JsonPatchOperation[] = [];

  properties.forEach((property: IPropertyBag) => {
    patchBody.push({
      op: Operation.Add,
      path: `/${property.propertyName}`,
      value: `${property.propertyValue}`,
    });
  });

  tl.debug(JSON.stringify(patchBody));

  const customHeader = { Authorization: `Bearer ${getAuthToken()}` };

  const azdoWebApi = getWebApi(collectionUri);
  const jsonPatchBody: JsonPatchDocument[] = patchBody;

  try {
    tl.debug("Acquiring a build API object.");
    const buildApi = await azdoWebApi.getBuildApi();
    tl.debug("Creating a new build property with global Quality Gate Status");
    await buildApi.updateBuildProperties(customHeader, jsonPatchBody, teamProjectId, +buildId);
  } catch (exception) {
    tl.warning(
      "Failed to create a build property. Not blocking unless you are using the Sonar Pre-Deployment gate in Release Pipelines. Exception : " +
        exception
    );
  }
}

export function getWebApi(collectionUrl: string): vm.WebApi {
  const accessToken = getAuthToken();
  const credentialHandler = vm.getBearerHandler(accessToken);
  return new vm.WebApi(collectionUrl, credentialHandler);
}

export function getAuthToken() {
  const auth = tl.getEndpointAuthorization("SYSTEMVSSCONNECTION", false);
  if (auth.scheme.toLowerCase() === "oauth") {
    return auth.parameters["AccessToken"];
  } else {
    throw new Error("Unable to get credential to perform rest API calls");
  }
}
