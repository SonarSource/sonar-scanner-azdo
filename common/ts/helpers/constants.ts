export const LATEST_DOC =
  "https://docs.sonarsource.com/sonarqube/latest/analyzing-source-code/scanners/sonarqube-extension-for-azure-devops/";
export const DEPRECATION_MESSAGE = `This task is deprecated. Please upgrade to the latest version. For more information, refer to ${LATEST_DOC}`;

export enum Status {
  OK = 200,
}
