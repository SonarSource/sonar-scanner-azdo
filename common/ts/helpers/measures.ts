const HOURS_IN_DAY = 8;

interface Formatter {
  (value: string | number, options?: any): string;
}

export function formatMeasure(
  value: string | number | undefined,
  type: string,
  options?: any
): string {
  const formatter = getFormatter(type);
  return useFormatter(value, formatter, options);
}

function useFormatter(
  value: string | number | undefined,
  formatter: Formatter,
  options?: any
): string {
  return value !== undefined && value !== "" ? formatter(value, options) : "";
}

function getFormatter(type: string): Formatter {
  const FORMATTERS: { [type: string]: Formatter } = {
    INT: intFormatter,
    SHORT_INT: shortIntFormatter,
    FLOAT: floatFormatter,
    PERCENT: percentFormatter,
    WORK_DUR: durationFormatter,
    SHORT_WORK_DUR: shortDurationFormatter,
    RATING: ratingFormatter,
    LEVEL: levelFormatter,
    MILLISEC: millisecondsFormatter,
    COMPARATOR: comparatorFormatter,
  };
  return FORMATTERS[type] || noFormatter;
}

function numberFormatter(
  value: number,
  minimumFractionDigits = 0,
  maximumFractionDigits = minimumFractionDigits
) {
  const { format } = new Intl.NumberFormat("en-US", {
    minimumFractionDigits,
    maximumFractionDigits,
  });
  return format(value);
}

function noFormatter(value: string | number): string {
  return value.toString();
}

function intFormatter(value: number): string {
  return numberFormatter(value);
}

function shortIntFormatter(value: number): string {
  if (value >= 1e9) {
    return numberFormatter(value / 1e9) + "G";
  } else if (value >= 1e6) {
    return numberFormatter(value / 1e6) + "M";
  } else if (value >= 1e4) {
    return numberFormatter(value / 1e3) + "k";
  } else if (value >= 1e3) {
    return numberFormatter(value / 1e3, 0, 1) + "k";
  } else {
    return numberFormatter(value);
  }
}

function floatFormatter(value: number): string {
  return numberFormatter(value, 1, 5);
}

function percentFormatter(value: string | number, options: { decimals?: number } = {}): string {
  if (typeof value === "string") {
    value = parseFloat(value);
  }
  if (options.decimals) {
    return numberFormatter(value, options.decimals) + "%";
  }
  return value === 100 ? "100%" : numberFormatter(value, 1) + "%";
}

function ratingFormatter(value: string | number): string {
  if (typeof value === "string") {
    value = parseInt(value, 10);
  }
  return String.fromCharCode(97 + value - 1).toUpperCase();
}

function levelFormatter(value: string): string {
  const l10nKeys = {
    ERROR: "Failed",
    WARN: "Warning",
    OK: "Passed",
    NONE: "None",
  };
  const result = l10nKeys[value.toUpperCase()];
  return result ? result : value;
}

function comparatorFormatter(value: string): string {
  const l10nKeys = {
    EQ: "&#61;",
    GT: "&#62;",
    LT: "&#60;",
    NE: "&#8800;",
  };
  const result = l10nKeys[value.toUpperCase()];
  return result ? result : value;
}

function millisecondsFormatter(value: number): string {
  const ONE_SECOND = 1000;
  const ONE_MINUTE = 60 * ONE_SECOND;
  if (value >= ONE_MINUTE) {
    const minutes = Math.round(value / ONE_MINUTE);
    return `${minutes}min`;
  } else if (value >= ONE_SECOND) {
    const seconds = Math.round(value / ONE_SECOND);
    return `${seconds}s`;
  } else {
    return `${value}ms`;
  }
}

function shouldDisplayDays(days: number): boolean {
  return days > 0;
}

function shouldDisplayDaysInShortFormat(days: number): boolean {
  return days > 0.9;
}

function shouldDisplayHours(days: number, hours: number): boolean {
  return hours > 0 && days < 10;
}

function shouldDisplayHoursInShortFormat(hours: number): boolean {
  return hours > 0.9;
}

function shouldDisplayMinutes(days: number, hours: number, minutes: number): boolean {
  return minutes > 0 && hours < 10 && days === 0;
}

function addSpaceIfNeeded(value: string): string {
  return value.length > 0 ? value + " " : value;
}

function formatDuration(isNegative: boolean, days: number, hours: number, minutes: number): string {
  let formatted = "";
  if (shouldDisplayDays(days)) {
    formatted += (isNegative ? -1 * days : days) + "d";
  }
  if (shouldDisplayHours(days, hours)) {
    formatted = addSpaceIfNeeded(formatted);
    formatted += (isNegative && formatted.length === 0 ? -1 * hours : hours) + "h";
  }
  if (shouldDisplayMinutes(days, hours, minutes)) {
    formatted = addSpaceIfNeeded(formatted);
    formatted += (isNegative && formatted.length === 0 ? -1 * minutes : minutes) + "min";
  }
  return formatted;
}

function formatDurationShort(
  isNegative: boolean,
  days: number,
  hours: number,
  minutes: number
): string {
  if (shouldDisplayDaysInShortFormat(days)) {
    const roundedDays = Math.round(days);
    const formattedDays = formatMeasure(isNegative ? -1 * roundedDays : roundedDays, "SHORT_INT");
    return formattedDays + "d";
  }

  if (shouldDisplayHoursInShortFormat(hours)) {
    const roundedHours = Math.round(hours);
    const formattedHours = formatMeasure(
      isNegative ? -1 * roundedHours : roundedHours,
      "SHORT_INT"
    );
    return formattedHours + "h";
  }

  const formattedMinutes = formatMeasure(isNegative ? -1 * minutes : minutes, "SHORT_INT");
  return formattedMinutes + "min";
}

function durationFormatter(value: string | number): string {
  if (typeof value === "string") {
    value = parseInt(value, 10);
  }
  if (value === 0) {
    return "0";
  }
  const hoursInDay = HOURS_IN_DAY;
  const isNegative = value < 0;
  const absValue = Math.abs(value);
  const days = Math.floor(absValue / hoursInDay / 60);
  let remainingValue = absValue - days * hoursInDay * 60;
  const hours = Math.floor(remainingValue / 60);
  remainingValue -= hours * 60;
  return formatDuration(isNegative, days, hours, remainingValue);
}

function shortDurationFormatter(value: string | number): string {
  if (typeof value === "string") {
    value = parseInt(value, 10);
  }
  if (value === 0) {
    return "0";
  }
  const hoursInDay = HOURS_IN_DAY;
  const isNegative = value < 0;
  const absValue = Math.abs(value);
  const days = absValue / hoursInDay / 60;
  let remainingValue = absValue - Math.floor(days) * hoursInDay * 60;
  const hours = remainingValue / 60;
  remainingValue -= Math.floor(hours) * 60;
  return formatDurationShort(isNegative, days, hours, remainingValue);
}
