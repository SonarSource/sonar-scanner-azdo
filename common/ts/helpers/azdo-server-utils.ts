import * as path from "path";
import * as fs from "fs-extra";
import * as tl from "azure-pipelines-task-lib/task";
import * as azdoApiUtils from "./../helpers/azdo-api-utils";

export function publishBuildSummary(summary: string, endpointType = "SonarQube") {
  uploadBuildSummary(saveBuildSummary(summary), `${endpointType} Analysis Report`);
}

export function saveBuildSummary(summary: string): string {
  const filePath = path.join(getStagingDirectory(), "SonarQubeBuildSummary.md");
  fs.writeFileSync(filePath, summary);
  tl.debug(`[SQ] Summary saved at: ${filePath}`);
  return filePath;
}

export function getStagingDirectory(): string {
  const dir = path.join(tl.getVariable("build.artifactStagingDirectory"), ".sqAnalysis");
  fs.ensureDirSync(dir);
  return dir;
}

export function uploadBuildSummary(summaryPath: string, title: string): void {
  tl.debug(`[SQ] Uploading build summary from ${summaryPath}`);
  tl.command(
    "task.addattachment",
    {
      type: "Distributedtask.Core.Summary",
      name: title,
    },
    summaryPath
  );
}

export async function fillBuildProperty(sonarQualityGateStatus: string) {
  const properties: azdoApiUtils.IPropertyBag[] = [];

  properties.push({
    propertyName: "sonarglobalqualitygate",
    propertyValue: sonarQualityGateStatus,
  });

  await azdoApiUtils.addBuildProperty(properties);
}
