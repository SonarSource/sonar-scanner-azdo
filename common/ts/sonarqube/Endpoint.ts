import * as tl from "azure-pipelines-task-lib/task";
import { HttpProxyAgent } from "http-proxy-agent";
import { HttpsProxyAgent } from "https-proxy-agent";
import { RequestInit } from "node-fetch";
import { URL } from "url";
import { getProxyFromURI } from "../helpers/proxyFromEnv";
import { PROP_NAMES } from "../helpers/utils";

export const REQUEST_TIMEOUT = 60000;

export enum EndpointType {
  SonarCloud = "SonarCloud",
  SonarQube = "SonarQube",
}

export interface EndpointData {
  url: string;
  token?: string;
  username?: string;
  password?: string;
  organization?: string;
}

export default class Endpoint {
  constructor(
    public type: EndpointType,
    private readonly data: EndpointData,
  ) {}

  public get auth() {
    if (!this.data.token && this.data.password && this.data.password.length > 0) {
      return { user: this.data.username, pass: this.data.password };
    }
    return { user: this.data.token || this.data.username };
  }

  public get organization() {
    return this.data.organization;
  }

  public get url() {
    return this.data.url;
  }

  toFetchOptions(endpointUrl: string): Partial<RequestInit> {
    const options: Partial<RequestInit> = {
      method: "get",
      timeout: REQUEST_TIMEOUT,
    };

    // Add HTTP auth from this.auth
    options.headers = {
      Authorization:
        "Basic " + Buffer.from(`${this.auth.user}:${this.auth.pass ?? ""}`).toString("base64"),
    };

    // Add proxy configuration, when relevant
    const envProxyUrl = getProxyFromURI(new URL(endpointUrl));
    const azureProxyUrl = tl.getHttpProxyConfiguration()?.proxyFormattedUrl;
    const ProxyAgentClass = endpointUrl.startsWith("https://") ? HttpsProxyAgent : HttpProxyAgent;
    if (envProxyUrl) {
      options.agent = new ProxyAgentClass(envProxyUrl);
      tl.debug("Using proxy agent from environment: " + JSON.stringify(options.agent));
    } else if (azureProxyUrl) {
      options.agent = new ProxyAgentClass(azureProxyUrl);
      tl.debug("Using proxy agent from Azure: " + JSON.stringify(options.agent));
    } else {
      tl.debug("Not using a proxy agent");
    }

    return options;
  }

  public toJson() {
    return JSON.stringify({ type: this.type, data: this.data });
  }

  public toSonarProps() {
    return {
      [PROP_NAMES.HOST_URL]: this.data.url,
      [PROP_NAMES.LOGIN]: this.data.token || this.data.username,
      [PROP_NAMES.PASSSWORD]:
        this.data.password && this.data.password.length > 0 ? this.data.password : null,
      [PROP_NAMES.ORG]: this.data.organization,
    };
  }

  public static getEndpoint(id: string, type: EndpointType): Endpoint {
    const url = tl.getEndpointUrl(id, false);
    const token = tl.getEndpointAuthorizationParameter(
      id,
      "apitoken",
      type !== EndpointType.SonarCloud,
    );
    const username = tl.getEndpointAuthorizationParameter(id, "username", true);
    const password = tl.getEndpointAuthorizationParameter(id, "password", true);
    const organization = tl.getInput("organization", type === EndpointType.SonarCloud);
    return new Endpoint(type, { url, token, username, password, organization });
  }
}
