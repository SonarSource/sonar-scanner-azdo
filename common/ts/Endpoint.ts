/* eslint-disable import/newline-after-import */
import * as tl from 'vsts-task-lib/task';
import * as request from 'request';
import { PROP_NAMES } from './utils';

export enum EndpointType {
  SonarCloud = 'SonarCloud',
  SonarQube = 'SonarQube'
}

export interface EndpointData {
  url: string;
  token?: string;
  username?: string;
  password?: string;
  organization?: string;
}

export interface RequestData {
  [x: string]: any;
}

export default class Endpoint {
  constructor(public type: EndpointType, public data: EndpointData) {}

  public toJson() {
    return JSON.stringify({ type: this.type, data: this.data });
  }

  public toSonarProps() {
    return {
      [PROP_NAMES.HOST_URL]: this.data.url,
      [PROP_NAMES.LOGIN]: this.data.token || this.data.username,
      [PROP_NAMES.PASSSWORD]: this.data.password,
      [PROP_NAMES.ORG]: this.data.organization
    };
  }

  public apiGetJSON(path: string, query?: RequestData): Promise<any> {
    tl.debug(`[SQ] API GET: '${path}' with query "${JSON.stringify(query)}"`);
    return new Promise((resolve, reject) => {
      const options: request.CoreOptions = {};
      if (this.data.token) {
        options.auth = { user: this.data.token };
      } else {
        options.auth = { user: this.data.username, pass: this.data.password };
      }
      if (query) {
        options.qs = query;
        options.useQuerystring = true;
      }
      request.get(
        {
          method: 'GET',
          baseUrl: this.data.url,
          uri: path,
          json: true,
          ...options
        },
        (error, response, body) => {
          if (error) {
            return reject(`[SQ] API GET '${path}' failed, error was: ${error}`);
          }

          if (response.statusCode < 200 || response.statusCode >= 300) {
            return reject(
              new Error(`[SQ] API GET '${path}' failed, status code was: ${response.statusCode}`)
            );
          }

          return resolve(body || {});
        }
      );
    });
  }

  public static getEndpoint(id: string, type: EndpointType): Endpoint {
    const url = tl.getEndpointUrl(id, false);
    const token = tl.getEndpointAuthorizationParameter(
      id,
      'apitoken',
      type === EndpointType.SonarQube
    );
    const username = tl.getEndpointAuthorizationParameter(id, 'username', true);
    const password = tl.getEndpointAuthorizationParameter(id, 'password', true);
    const organization = tl.getInput('organization', true);
    return new Endpoint(type, { url, token, username, password, organization });
  }
}
