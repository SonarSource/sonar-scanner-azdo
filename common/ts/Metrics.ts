import * as tl from 'vsts-task-lib/task';
import Endpoint from './Endpoint';

interface IMetric {
  custom?: boolean;
  decimalScale?: number;
  description?: string;
  direction?: number;
  domain?: string;
  hidden?: boolean;
  key: string;
  name: string;
  qualitative?: boolean;
  type: string;
}

interface MetricsResponse {
  metrics: IMetric[];
  p: number;
  ps: number;
  total: number;
}

export default class Metrics {
  constructor(public metrics: IMetric[]) {}

  public getMetricByKey(key: string) {
    return this.metrics.find(metric => metric.key === key);
  }

  public static getAllMetrics(endpoint: Endpoint): Promise<Metrics> {
    return inner().catch(err => {
      if (err && err.message) {
        tl.debug(err.message);
      }
      throw new Error(`[SQ] Could not fetch metrics`);
    });

    function inner(
      data: { f?: string; p?: number; ps?: number } = { f: 'name', ps: 500 },
      prev?: MetricsResponse
    ): Promise<Metrics> {
      return endpoint.apiGetJSON('/api/metrics/search', data).then((r: MetricsResponse) => {
        const result = prev ? prev.metrics.concat(r.metrics) : r.metrics;
        if (r.p * r.ps >= r.total) {
          return new Metrics(result);
        }
        return inner({ ...data, p: r.p + 1 }, { ...r, metrics: result });
      });
    }
  }
}
