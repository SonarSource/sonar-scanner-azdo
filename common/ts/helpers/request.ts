import * as request from 'request';
import * as semver from 'semver';
import * as tl from 'vsts-task-lib/task';
import Endpoint from '../sonarqube/Endpoint';

interface RequestData {
  [x: string]: any;
}

function get(endpoint: Endpoint, path: string, isJson: boolean, query?: RequestData): Promise<any> {
  tl.debug(`[SQ] API GET: '${path}' with query "${JSON.stringify(query)}"`);
  return new Promise((resolve, reject) => {
    const options: request.CoreOptions = {
      auth: endpoint.auth
    };
    if (query) {
      options.qs = query;
      options.useQuerystring = true;
    }
    request.get(
      {
        method: 'GET',
        baseUrl: endpoint.url,
        uri: path,
        json: isJson,
        ...options
      },
      (error, response, body) => {
        if (error) {
          return logAndReject(
            reject,
            `[SQ] API GET '${path}' failed, error was: ${JSON.stringify(error)}`
          );
        }
        tl.debug(
          `Response: ${response.statusCode} Body: "${isString(body) ? body : JSON.stringify(body)}"`
        );
        if (response.statusCode < 200 || response.statusCode >= 300) {
          return logAndReject(
            reject,
            `[SQ] API GET '${path}' failed, status code was: ${response.statusCode}`
          );
        }
        return resolve(body || (isJson ? {} : ''));
      }
    );
  });
}

function isString(x) {
  return Object.prototype.toString.call(x) === '[object String]';
}

export function getJSON(endpoint: Endpoint, path: string, query?: RequestData): Promise<any> {
  return get(endpoint, path, true, query);
}

export function getServerVersion(endpoint: Endpoint): Promise<semver.SemVer> {
  return get(endpoint, '/api/server/version', false).then(semver.coerce);
}

function logAndReject(reject, errMsg) {
  tl.debug(errMsg);
  return reject(new Error(errMsg));
}
