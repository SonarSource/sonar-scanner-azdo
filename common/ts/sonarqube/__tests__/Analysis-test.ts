import Analysis from "../Analysis";
import { getJSON } from "../../helpers/request";
import Metrics from "../Metrics";
import Endpoint, { EndpointType } from "../Endpoint";

jest.mock("../../helpers/request", () => ({
  getJSON: jest.fn(() =>
    Promise.resolve({
      projectStatus: {
        status: "ERROR",
        conditions: [
          {
            status: "ERROR",
            metricKey: "bugs",
            comparator: "GT",
            errorThreshold: "0",
            actualValue: "1",
          },
        ],
      },
    })
  ),
}));

jest.mock("azure-pipelines-task-lib/task", () => ({
  debug: jest.fn(),
  error: jest.fn(),
}));

const METRICS = new Metrics([{ key: "bugs", name: "Bugs", type: "INT" }]);
const ENDPOINT = new Endpoint(EndpointType.SonarQube, { url: "https://endpoint.url" });
const GET_ANALYSIS_DATA = {
  analysisId: "analysisId",
  dashboardUrl: "https://dashboard.url",
  endpoint: ENDPOINT,
  metrics: METRICS,
};

beforeEach(() => {
  (getJSON as jest.Mock<any>).mockClear();
});

it("should generate an analysis status with error", async () => {
  const analysis = await Analysis.getAnalysis(GET_ANALYSIS_DATA);
  expect(getJSON).toHaveBeenCalledWith(ENDPOINT, "/api/qualitygates/project_status", {
    analysisId: "analysisId",
  });
  expect(analysis.status).toBe("ERROR");
  expect(analysis.getFailedConditions()).toHaveLength(1);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should generate a green analysis status", async () => {
  (getJSON as jest.Mock<any>).mockImplementationOnce(() =>
    Promise.resolve({ projectStatus: { status: "SUCCESS", conditions: [] } })
  );

  const analysis = await Analysis.getAnalysis(GET_ANALYSIS_DATA);
  expect(getJSON).toHaveBeenCalledWith(ENDPOINT, "/api/qualitygates/project_status", {
    analysisId: "analysisId",
  });
  expect(analysis.status).toBe("SUCCESS");
  expect(analysis.getFailedConditions()).toHaveLength(0);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should not fail when metrics are missing", async () => {
  const analysis = await Analysis.getAnalysis({
    ...GET_ANALYSIS_DATA,
    dashboardUrl: undefined,
    metrics: undefined,
  });
  expect(getJSON).toHaveBeenCalledWith(ENDPOINT, "/api/qualitygates/project_status", {
    analysisId: "analysisId",
  });
  expect(analysis.status).toBe("ERROR");
  expect(analysis.getFailedConditions()).toHaveLength(1);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should display the project name", async () => {
  const analysis = await Analysis.getAnalysis({
    ...GET_ANALYSIS_DATA,
    projectName: "project_name",
  });
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});
