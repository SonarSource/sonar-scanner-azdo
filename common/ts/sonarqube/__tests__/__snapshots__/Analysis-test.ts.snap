// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the project name 1`] = `
"<div style="padding-top: 8px;">
      <span>project_name Quality Gate</span>
      <span style="background-color: #d4333f;
      padding: 4px 12px;
      color: #fff;
      letter-spacing: 0.02em;
      line-height: 24px;
      font-weight: 600;
      font-size: 12px;
      margin-left: 15px;">
        Failed
      </span>
    </div> 
<table border="0" style="
      margin-top: 8px;
      border-top: 1px solid #eee;
      border-collapse: separate;
      border-spacing: 0 4px;
    ">
      <tbody>
        <tr>
        <td><span style="padding-right:4px;">Bugs</span></td>
        <td style="text-align:center; color:#fff; background-color:#d4333f;">
          <span style="padding:0px 4px; line-height:20px;">1</span>
        </td>
        <td>
          <span style="padding-left:4px">&#62;</span>
          <span style="padding-left:4px">0</span>
        </td>
      </tr>
      </tbody>
    </table>

 
[Detailed SonarQube report &gt;](https://dashboard.url)"
`;

exports[`should generate a green analysis status 1`] = `
"<div style="padding-top: 8px;">
      <span>Quality Gate</span>
      <span style="background-color: #b4b4b4;
      padding: 4px 12px;
      color: #fff;
      letter-spacing: 0.02em;
      line-height: 24px;
      font-weight: 600;
      font-size: 12px;
      margin-left: 15px;">
        SUCCESS
      </span>
    </div> 
 
[Detailed SonarQube report &gt;](https://dashboard.url)"
`;

exports[`should generate an analysis status with error 1`] = `
"<div style="padding-top: 8px;">
      <span>Quality Gate</span>
      <span style="background-color: #d4333f;
      padding: 4px 12px;
      color: #fff;
      letter-spacing: 0.02em;
      line-height: 24px;
      font-weight: 600;
      font-size: 12px;
      margin-left: 15px;">
        Failed
      </span>
    </div> 
<table border="0" style="
      margin-top: 8px;
      border-top: 1px solid #eee;
      border-collapse: separate;
      border-spacing: 0 4px;
    ">
      <tbody>
        <tr>
        <td><span style="padding-right:4px;">Bugs</span></td>
        <td style="text-align:center; color:#fff; background-color:#d4333f;">
          <span style="padding:0px 4px; line-height:20px;">1</span>
        </td>
        <td>
          <span style="padding-left:4px">&#62;</span>
          <span style="padding-left:4px">0</span>
        </td>
      </tr>
      </tbody>
    </table>

 
[Detailed SonarQube report &gt;](https://dashboard.url)"
`;

exports[`should not fail when metrics are missing 1`] = `
"<div style="padding-top: 8px;">
      <span>Quality Gate</span>
      <span style="background-color: #d4333f;
      padding: 4px 12px;
      color: #fff;
      letter-spacing: 0.02em;
      line-height: 24px;
      font-weight: 600;
      font-size: 12px;
      margin-left: 15px;">
        Failed
      </span>
    </div>"
`;
